applications/serial_lte_modem/src/gnss/slm_at_gnss.c

```diff
@@ -0,0 +1,832 @@
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr.h>
+#include <stdio.h>
+#include <logging/log.h>
+#include <date_time.h>
+#include <pm_config.h>
+#include <net/cloud.h>
+#include <nrf_modem_gnss.h>
+#include <net/nrf_cloud_agps.h>
+#include <net/nrf_cloud_pgps.h>
+#include <net/nrf_cloud_cell_pos.h>
+#include "slm_util.h"
+#include "slm_at_host.h"
+#include "slm_at_gnss.h"
+
+LOG_MODULE_REGISTER(slm_gnss, CONFIG_SLM_LOG_LEVEL);
+
+#define SERVICE_INFO_GPS \
+	"{\"state\":{\"reported\":{\"device\": {\"serviceInfo\":{\"ui\":[\"GPS\"]}}}}}"
+
+/**@brief GNSS operations. */
+enum slm_gnss_operation {
+	GPS_STOP,
+	GPS_START,
+	nRF_CLOUD_DISCONNECT = GPS_STOP,
+	nRF_CLOUD_CONNECT    = GPS_START,
+	AGPS_STOP            = GPS_STOP,
+	AGPS_START           = GPS_START,
+	PGPS_STOP            = GPS_STOP,
+	PGPS_START           = GPS_START,
+	CELLPOS_STOP         = GPS_STOP,
+	CELLPOS_START_SCELL  = GPS_START,
+	CELLPOS_START_MCELL
+};
+
+static struct k_work agps_req;
+static struct k_work pgps_req;
+static struct k_work cell_pos_req;
+static enum nrf_cloud_cell_pos_type cell_pos_type;
+
+static struct cloud_backend *nrf_cloud;
+static bool nrf_cloud_ready;
+static bool location_signify;
+static uint64_t ttft_start;
+static enum {
+	RUN_TYPE_NONE,
+	RUN_TYPE_GPS,
+	RUN_TYPE_AGPS,
+	RUN_TYPE_PGPS,
+	RUN_TYPE_CELL_POS
+} run_type;
+
+K_SEM_DEFINE(sem_date_time, 0, 1);
+
+/* global functions defined in different files */
+void rsp_send(const uint8_t *str, size_t len);
+
+/* global variable defined in different files */
+extern struct k_work_q slm_work_q;
+extern struct at_param_list at_param_list;
+extern char rsp_buf[CONFIG_SLM_SOCKET_RX_MAX * 2];
+
+static int read_agps_req(struct gps_agps_request *req)
+{
+	int err;
+	struct nrf_modem_gnss_agps_data_frame agps_data;
+
+	err = nrf_modem_gnss_read((void *)&agps_data, sizeof(agps_data),
+					NRF_MODEM_GNSS_DATA_AGPS_REQ);
+	if (err) {
+		LOG_ERR("Failed to read GNSS AGPS req, error %d", err);
+		return -EAGAIN;
+	}
+
+	req->sv_mask_ephe = agps_data.sv_mask_ephe,
+	req->sv_mask_alm  = agps_data.sv_mask_alm,
+	req->utc          = (agps_data.data_flags & NRF_MODEM_GNSS_AGPS_GPS_UTC_REQUEST) ? 1 : 0,
+	req->klobuchar    = (agps_data.data_flags & NRF_MODEM_GNSS_AGPS_KLOBUCHAR_REQUEST) ? 1 : 0,
+	req->nequick      = (agps_data.data_flags & NRF_MODEM_GNSS_AGPS_NEQUICK_REQUEST) ? 1 : 0,
+	req->system_time_tow =
+	      (agps_data.data_flags & NRF_MODEM_GNSS_AGPS_SYS_TIME_AND_SV_TOW_REQUEST) ? 1 : 0,
+	req->position     = (agps_data.data_flags & NRF_MODEM_GNSS_AGPS_POSITION_REQUEST) ? 1 : 0,
+	req->integrity    = (agps_data.data_flags & NRF_MODEM_GNSS_AGPS_INTEGRITY_REQUEST) ? 1 : 0;
+
+	return 0;
+}
+
+static void agps_req_wk(struct k_work *work)
+{
+	int err;
+	struct gps_agps_request req;
+
+	ARG_UNUSED(work);
+
+	err = read_agps_req(&req);
+	if (err) {
+		return;
+	}
+
+	err = nrf_cloud_agps_request(req);
+	if (err) {
+		LOG_ERR("Failed to request A-GPS data: %d", err);
+	}
```

Comment:
If you want so support using AGPS+PGPS at the same time, this function would need to check for run_type == RUN_TYPE_AGPS_PGS and then do this if true:
```
} else {
        if (!nrf_cloud_apgs_request_in_progress()) {
                /* AGPS data not expected, so move on to PGPS */
                nrf_cloud_pgps_notify_prediction();
        }
}

Reply:
Or, add a separate work function for this combined mode.